src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java

```diff
@@ -0,0 +1,367 @@
+            valueToConvert,
+            targetType.getSimpleName(),
+            failureMsg
+        );
+
+        if (recordProblem(ctxt, message,
+                ctxt.constructType(targetType), valueToConvert)) {
+            // Return sensible default based on target type
+            return getDefaultValue(targetType);
+        }
+
+        return NOT_HANDLED; // Limit reached
+    }
+
+    @Override
+    public Object handleWeirdNumberValue(DeserializationContext ctxt,
+            Class<?> targetType, Number valueToConvert, String failureMsg)
+            throws JacksonException {
+
+        String message = String.format(
+            "Cannot deserialize number %s to %s: %s",
+            valueToConvert,
+            targetType.getSimpleName(),
+            failureMsg
+        );
+
+        if (recordProblem(ctxt, message,
+                ctxt.constructType(targetType), valueToConvert)) {
+            return getDefaultValue(targetType);
+        }
+
+        return NOT_HANDLED;
+    }
+
+    @Override
+    public Object handleInstantiationProblem(DeserializationContext ctxt,
+            Class<?> instClass, Object argument, Throwable t)
+            throws JacksonException {
+
+        String message = String.format(
+            "Cannot instantiate %s: %s",
+            instClass.getSimpleName(),
+            t.getMessage()
+        );
+
+        if (recordProblem(ctxt, message,
+                ctxt.constructType(instClass), argument)) {
+            // Only return null if we can safely continue
+            // For some types, instantiation failure is fatal
+            if (canReturnNullFor(instClass)) {
+                return null;
+            }
+        }
+
+        return NOT_HANDLED; // Cannot recover
+    }
+
+    /**
+     * Returns a sensible default value for the given type to allow
+     * deserialization to continue.
+     *
+     * <p>IMPORTANT: Only primitives get non-null defaults. Reference types
+     * (including boxed primitives) get null to avoid masking nullability issues.
+     */
+    private Object getDefaultValue(Class<?> type) {
+        // Primitives MUST have non-null defaults (cannot be null)
+        if (type == int.class) return 0;
+        if (type == long.class) return 0L;
+        if (type == double.class) return 0.0;
+        if (type == float.class) return 0.0f;
+        if (type == boolean.class) return false;
+        if (type == byte.class) return (byte) 0;
+        if (type == short.class) return (short) 0;
+        if (type == char.class) return '\0';
+
+        // Reference types (including Integer, Long, etc.) get null
+        // This avoids masking nullability issues in the domain model
+        return null;
+    }
+
+    /**
+     * Checks if it's safe to return null for a given type after
+     * instantiation failure.
+     */
+    private boolean canReturnNullFor(Class<?> type) {
+        // Cannot return null for primitives or arrays
```

Comment:
Why not for Arrays? `null` is valid for them, similar to `Collection`s etc (and unlike primitives)